src/Engine/ProtoCore/DSASM/HeapObjects.cs

```diff
@@ -0,0 +1,659 @@
+        }
+
+        /// <summary>
+        /// Copy an array and coerce its elements/values to target type
+        /// </summary>
+        /// <param name="array"></param>
+        /// <param name="type"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue CopyArray(Type type, RuntimeCore runtimeCore)
+        {
+            StackValue[] elements = new StackValue[VisibleSize];
+            for (int i = 0; i < VisibleSize; i++)
+            {
+                StackValue coercedValue = TypeSystem.Coerce(GetItemAt(i), type, runtimeCore);
+                elements[i] = coercedValue;
+            }
+
+            Dictionary<StackValue, StackValue> dict = null;
+            if (Dict != null)
+            {
+                dict = new Dictionary<StackValue, StackValue>(new StackValueComparer(runtimeCore));
+                foreach (var pair in Dict)
+                {
+                    StackValue key = pair.Key;
+                    StackValue value = pair.Value;
+                    StackValue coercedValue = TypeSystem.Coerce(value, type, runtimeCore);
+
+                    dict[key] = coercedValue;
+                }
+            }
+
+            var svArray = heap.AllocateArray(elements);
+            var array = heap.ToHeapObject<DSArray>(svArray);
+            array.Dict = dict;
+            return svArray;
+        }
+
+        /// <summary>
+        /// array[index] = value. The array will be expanded if necessary.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue SetValueForIndex(int index, StackValue value, RuntimeCore runtimeCore)
+        {
+            index = ExpandByAcessingAt(index);
+            if (index < 0)
+                index += VisibleSize;
+
+            if (index >= VisibleSize || index < 0)
+            {
+                if (runtimeCore != null)
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kOverIndexing, Resources.kArrayOverIndexed);
+                return StackValue.Null;
+            }
+
+            StackValue oldValue = GetItemAt(index);
+            SetItemAt(index, value);
+            return oldValue;
+        }
+
+        /// <summary>
+        /// array[index] = value. Here index can be any type. 
+        /// 
+        /// Note this function doesn't support the replication of array indexing.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue SetValueForIndex(StackValue index, StackValue value, RuntimeCore runtimeCore)
+        {
+            if (index.IsNumeric)
+            {
+                index = index.ToInteger();
+                return SetValueForIndex((int)index.opdata, value, runtimeCore);
```

Comment:
Do you think `AddressType.Double` should even be allowed here (I presume that's the reason you did a `ToInteger`). This is probably a question for later because if we update the first argument for `SetValueForIndex` to `int`, the contract may be clearer.